marlowe-playground-client/src/Reachability.purs

```diff
@@ -166,38 +143,128 @@ nullifyAsserts (Let valId val cont) = Let valId val (nullifyAsserts cont)
 
 nullifyAsserts (Assert obs cont) = Assert TrueObs cont
 
-generateSubproblems :: Contract -> List (Unit -> ContractPath /\ Contract)
-generateSubproblems contract = reverse (foldBreadthContractWithZipper generateSubproblem Nil (nullifyAsserts contract))
-
 startReachabilityAnalysis ::
   forall m.
   MonadAff m =>
   SPSettings_ SPParams_ ->
   Contract ->
   S.State -> HalogenM State Action ChildSlots Void m ReachabilityAnalysisData
 startReachabilityAnalysis settings contract state = do
-  case subproblemList of
-    Cons head tail -> do
-      let
-        path /\ subcontract = head unit
+  case getNextSubproblem isValidSubproblem initialSubproblems Nil of
+    Nothing -> pure AllReachable
+    Just ((contractZipper /\ subcontract /\ newChildren) /\ newSubproblems) -> do
       let
+        numSubproblems = countSubproblems isValidSubproblem (newChildren <> newSubproblems)
+
+        newPath /\ newContract = expandSubproblem contractZipper
+
         progress =
           ( InProgress
-              { currPath: path
-              , currContract: subcontract
+              { currPath: newPath
+              , currContract: newContract
+              , currChildren: newChildren
               , originalState: state
-              , subproblems: tail
-              , numSubproblems: 1 + length tail
+              , subproblems: newSubproblems
+              , numSubproblems: 1 + numSubproblems
               , numSolvedSubproblems: 0
+              , unreachableSubcontracts: Nil
               }
           )
       assign _analysisState (ReachabilityAnalysis progress)
       response <- checkContractForReachability settings subcontract state
       result <- updateWithResponse settings progress response
       pure result
-    Nil -> pure AllReachable
   where
-  subproblemList = generateSubproblems contract
+  initialSubproblems = initSubproblems contract
+
+initSubproblems :: Contract -> RemainingSubProblemInfo
+initSubproblems c = Cons (HeadZip /\ c) Nil
+
+countFix :: forall a. (a -> Maybe a) -> a -> Int
+countFix fun a0 = countFix_tailrec fun a0 0
```

Comment:
can't you use `foldl` for this?

Reply:
I could not think of a way, since there is no structure to traverse. I did a quick search in Pursuit and couldn't find anything similar either